README.rst
IMU Brick
=========

This repository contains the firmware source code and the hardware design
files. The documentation generator configs can be found at
https://github.com/Tinkerforge/generators

Repository Content
------------------

software/:
 * examples/: Examples for all supported languages
 * build/: Compiled files
 * src/: Source code of firmware
 * Makefile: Makefile to build project

hardware/:
 * Contains KiCad project files and additionally schematics as PDF

datasheets/:
 * Contains datasheets for sensors and complex ICs that are used

Hardware
--------

The hardware is designed with the open source EDA Suite KiCad
(http://www.kicad-pcb.org). Before you are able to open the files,
you have to install the Tinkerforge kicad-libraries
(https://github.com/Tinkerforge/kicad-libraries). You can either clone
them directly in hardware/ or clone them in a separate folder and
symlink them into hardware/
(ln -s kicad_path/kicad-libraries project_path/hardware). After that you
can open the .pro file in hardware/ with KiCad and from there view and
modify the schematics and the PCB layout.

Software
--------

If you want to do your own Brick/Bricklet firmware development we highly
recommend that you use our build environment setup script and read the
tutorial (https://www.tinkerforge.com/en/doc/Tutorials/Tutorial_Build_Environment/Tutorial.html).

To compile the C code we recommend you to install the newest GNU Arm Embedded
Toolchain (https://launchpad.net/gcc-arm-embedded/+download).
You also need to install bricklib (https://github.com/Tinkerforge/bricklib).

You can either clone it directly in software/src/ or clone it in a
separate folder and symlink it into software/src/
(ln -s bricklib_path/bricklib project_path/software/src/). Finally make sure to
have CMake installed (http://www.cmake.org/cmake/resources/software.html).

After that you can build the firmware by invoking make in software/.
The firmware (.bin) can then be found in software/build/ and uploaded
with brickv (click button "Flashing" on start screen).


software/changelog
2012-02-15: 1.0.0 (ca8e321)
- Initial version

2012-03-02: 1.0.1 (9fd6a77)
- Use quaternion instead of magnetometer for direction LEDs

2012-03-05: 1.0.2 (5954fed)
- Fix bug with timeouts on Windows (through bricklib)

2012-03-11: 1.0.3 (4a6976a)
- Fix bug with ADC calibration readout (through bricklib)

2012-03-29: 1.0.4 (ab66907)
- Remove catching up of list signal messages (for chibi)

2012-06-03: 1.0.5 (c9d7623)
- Fix wrong defines (see issue #1 on GitHub)

2012-06-06: 1.0.6 (9ac5a64)
- Use better compass approximation

2012-06-30: 1.0.7 (8704fda)
- Add get_chip_temperature and reset function

2012-07-02: 1.0.8 (05820a1)
- Add USB hotplug support (through bricklib)

2012-09-20: 1.0.9 (2c1d6e6)
- Set interrupt pins to input default (otherwise they can interfere with WIFI)

2012-10-26: 1.0.10 (a3e8b0f)
- Change Bricklet Context size from 250 to 256 (through bricklib)

2013-01-07: 1.0.11 (ae19568)
- Return correctly from interrupt (fixes bug related to GPS Bricklet)

2013-01-17: 2.0.0 (a643fda)
- Support for Protocol 2.0

2013-04-11: 2.0.1 (f82f568)
- Add watchdog timer (restart after 16s of "deadlock")

2013-05-13: 2.0.2 (a004269)
- Add orientation calculation on/off API, make I2C communication synchronous

2013-06-27: 2.0.3 (d158a14)
- Use tick_task to detect USB connect/disconnect instead of interrupt
  (through bricklib)

2014-07-07: 2.0.4 (d3e07ee)
- Remove USB hotplug support (through bricklib)

2014-11-25: 2.3.0 (41bfb2f)
- Add support for new SPI protocol

2015-07-27: 2.3.1 (92bb37e)
- Implement set/get acceleration/magnetometer range

2016-08-29: 2.3.2 (442b1c7)
- Make sure to use all chunks when writing Bricklet plugin (through bricklib)

2017-05-11: 2.3.3 (959840d)
- Add support for Co-MCU Bricklets

2017-06-28: 2.3.4 (9f270d5)
- Fix incompatibility between Co-MCU support and LED Strip Bricklet (through bricklib)

2017-12-13: 2.3.5 (c4b5f3b)
- Make sure Brick does not restart on suspend/resume (through bricklib)
- Refactor initial enumeration handling, fix double-enumeration bugs (through bricklib)

2018-01-11: 2.3.6 (cf8d001)
- Fix message duplication bug (through bricklib)

2018-04-27: 2.3.7 (094e45b)
- Fix potential deadlock between USB stack and Co-MCU enumeration (through bricklib)

2018-07-31: 2.3.8 (13873b6)
- Add support for Isolator Bricklet (through bricklib)
- Improve Co-MCU Bricklet error handling (through bricklib)

2019-02-27: 2.3.9 (0168e2f)
- Changes through bricklib:
  - Fix Bricklet port miss-identification
  - Increase SPI stack timeout (for RED Brick)
  - Fix bugs in set/get_spitfp_* functions


software/src/communication.h
/* imu-brick
 * Copyright (C) 2010-2012 Olaf Lüke <olaf@tinkerforge.com>
 *
 * communication.h: Implementation of IMU-Brick specific messages
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */


#ifndef COMMUNICATION_H
#define COMMUNICATION_H

#include <stdint.h>
#include <stdbool.h>

#include "bricklib/com/com_common.h"

#define FID_GET_ACCELERATION 1
#define FID_GET_MAGNETIC_FIELD 2
#define FID_GET_ANGULAR_VELOCITY 3
#define FID_GET_ALL_DATA 4
#define FID_GET_ORIENTATION 5
#define FID_GET_QUATERNION 6
#define FID_GET_IMU_TEMPERATURE 7
#define FID_LEDS_ON 8
#define FID_LEDS_OFF 9
#define FID_ARE_LEDS_ON 10
#define FID_SET_ACCELERATION_RANGE 11
#define FID_GET_ACCELERATION_RANGE 12
#define FID_SET_MAGNETOMETER_RANGE 13
#define FID_GET_MAGNETOMETER_RANGE 14
#define FID_SET_CONVERGENCE_SPEED 15
#define FID_GET_CONVERGENCE_SPEED 16
#define FID_SET_CALIBRATION 17
#define FID_GET_CALIBRATION 18
#define FID_SET_ACCELERATION_PERIOD 19
#define FID_GET_ACCELERATION_PERIOD 20
#define FID_SET_MAGNETIC_FIELD_PERIOD 21
#define FID_GET_MAGNETIC_FIELD_PERIOD 22
#define FID_SET_ANGULAR_VELOCITY_PERIOD 23
#define FID_GET_ANGULAR_VELOCITY_PERIOD 24
#define FID_SET_ALL_DATA_PERIOD 25
#define FID_GET_ALL_DATA_PERIOD 26
#define FID_SET_ORIENTATION_PERIOD 27
#define FID_GET_ORIENTATION_PERIOD 28
#define FID_SET_QUATERNION_PERIOD 29
#define FID_GET_QUATERNION_PERIOD 30
#define FID_ACCELERATION 31
#define FID_MAGNETIC_FIELD 32
#define FID_ANGULAR_VELOCITY 33
#define FID_ALL_DATA 34
#define FID_ORIENTATION 35
#define FID_QUATERNION 36
#define FID_ORIENTATION_CALCULATION_ON 37
#define FID_ORIENTATION_CALCULATION_OFF 38
#define FID_IS_ORIENTATION_CALCULATION_ON 39

#define COM_MESSAGE_USER_LAST_FID 39

#define COM_MESSAGES_USER \
	{FID_GET_ACCELERATION, (message_handler_func_t)get_acceleration}, \
	{FID_GET_MAGNETIC_FIELD, (message_handler_func_t)get_magnetic_field}, \
	{FID_GET_ANGULAR_VELOCITY, (message_handler_func_t)get_angular_velocity}, \
	{FID_GET_ALL_DATA, (message_handler_func_t)get_all_data}, \
	{FID_GET_ORIENTATION, (message_handler_func_t)get_orientation}, \
	{FID_GET_QUATERNION, (message_handler_func_t)get_quaternion}, \
	{FID_GET_IMU_TEMPERATURE, (message_handler_func_t)get_imu_temperature}, \
	{FID_LEDS_ON, (message_handler_func_t)leds_on}, \
	{FID_LEDS_OFF, (message_handler_func_t)leds_off}, \
	{FID_ARE_LEDS_ON, (message_handler_func_t)are_leds_on}, \
	{FID_SET_ACCELERATION_RANGE, (message_handler_func_t)set_acceleration_range}, \
	{FID_GET_ACCELERATION_RANGE, (message_handler_func_t)get_acceleration_range}, \
	{FID_SET_MAGNETOMETER_RANGE, (message_handler_func_t)set_magnetometer_range}, \
	{FID_GET_MAGNETOMETER_RANGE, (message_handler_func_t)get_magnetometer_range}, \
	{FID_SET_CONVERGENCE_SPEED, (message_handler_func_t)set_convergence_speed}, \
	{FID_GET_CONVERGENCE_SPEED, (message_handler_func_t)get_convergence_speed}, \
	{FID_SET_CALIBRATION, (message_handler_func_t)set_calibration}, \
	{FID_GET_CALIBRATION, (message_handler_func_t)get_calibration}, \
	{FID_SET_ACCELERATION_PERIOD, (message_handler_func_t)set_acceleration_period}, \
	{FID_GET_ACCELERATION_PERIOD, (message_handler_func_t)get_acceleration_period}, \
	{FID_SET_MAGNETIC_FIELD_PERIOD, (message_handler_func_t)set_magnetic_field_period}, \
	{FID_GET_MAGNETIC_FIELD_PERIOD, (message_handler_func_t)get_magnetic_field_period}, \
	{FID_SET_ANGULAR_VELOCITY_PERIOD, (message_handler_func_t)set_angular_velocity_period}, \
	{FID_GET_ANGULAR_VELOCITY_PERIOD, (message_handler_func_t)get_angular_velocity_period}, \
	{FID_SET_ALL_DATA_PERIOD, (message_handler_func_t)set_all_data_period}, \
	{FID_GET_ALL_DATA_PERIOD, (message_handler_func_t)get_all_data_period}, \
	{FID_SET_ORIENTATION_PERIOD, (message_handler_func_t)set_orientation_period}, \
	{FID_GET_ORIENTATION_PERIOD, (message_handler_func_t)get_orientation_period}, \
	{FID_SET_QUATERNION_PERIOD, (message_handler_func_t)set_quaternion_period}, \
	{FID_GET_QUATERNION_PERIOD, (message_handler_func_t)get_quaternion_period}, \
	{FID_ACCELERATION, (message_handler_func_t)NULL}, \
	{FID_MAGNETIC_FIELD, (message_handler_func_t)NULL}, \
	{FID_ANGULAR_VELOCITY, (message_handler_func_t)NULL}, \
	{FID_ALL_DATA, (message_handler_func_t)NULL}, \
	{FID_ORIENTATION, (message_handler_func_t)NULL}, \
	{FID_QUATERNION, (message_handler_func_t)NULL}, \
	{FID_ORIENTATION_CALCULATION_ON, (message_handler_func_t)orientation_calculation_on}, \
	{FID_ORIENTATION_CALCULATION_OFF, (message_handler_func_t)orientation_calculation_off}, \
	{FID_IS_ORIENTATION_CALCULATION_ON, (message_handler_func_t)is_orientation_calculation_on},


typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetAcceleration;

typedef struct {
	MessageHeader header;
	int16_t x;
	int16_t y;
	int16_t z;
} __attribute__((__packed__)) GetAccelerationReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetMagneticField;

typedef struct {
	MessageHeader header;
	int16_t x;
	int16_t y;
	int16_t z;
} __attribute__((__packed__)) GetMagneticFieldReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetAngularVelocity;

typedef struct {
	MessageHeader header;
	int16_t x;
	int16_t y;
	int16_t z;
} __attribute__((__packed__)) GetAngularVelocityReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetAllData;

typedef struct {
	MessageHeader header;
	int16_t acc_x;
	int16_t acc_y;
	int16_t acc_z;
	int16_t mag_x;
	int16_t mag_y;
	int16_t mag_z;
	int16_t ang_x;
	int16_t ang_y;
	int16_t ang_z;
	int16_t temperature;
} __attribute__((__packed__)) GetAllDataReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetOrientation;

typedef struct {
	MessageHeader header;
	int16_t roll;
	int16_t pitch;
	int16_t yaw;
} __attribute__((__packed__)) GetOrientationReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetQuaternion;

typedef struct {
	MessageHeader header;
	float x;
	float y;
	float z;
	float w;
} __attribute__((__packed__)) GetQuaternionReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetIMUTemperature;

typedef struct {
	MessageHeader header;
	int16_t temperature;
} __attribute__((__packed__)) GetIMUTemperatureReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) LedsOn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) LedsOff;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) AreLedsOn;

typedef struct {
	MessageHeader header;
	bool leds;
} __attribute__((__packed__)) AreLedsOnReturn;

typedef struct {
	MessageHeader header;
	uint8_t range;
} __attribute__((__packed__)) SetAccelerationRange;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetAccelerationRange;

typedef struct {
	MessageHeader header;
	uint8_t range;
} __attribute__((__packed__)) GetAccelerationRangeReturn;

typedef struct {
	MessageHeader header;
	uint8_t range;
} __attribute__((__packed__)) SetMagnetometerRange;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetMagnetometerRange;

typedef struct {
	MessageHeader header;
	uint8_t range;
} __attribute__((__packed__)) GetMagnetometerRangeReturn;

typedef struct {
	MessageHeader header;
	uint16_t speed;
} __attribute__((__packed__)) SetConvergenceSpeed;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetConvergenceSpeed;

typedef struct {
	MessageHeader header;
	uint16_t speed;
} __attribute__((__packed__)) GetConvergenceSpeedReturn;

typedef struct {
	MessageHeader header;
	uint8_t type_data;
	uint16_t data[10];
} __attribute__((__packed__)) SetCalibration;

typedef struct {
	MessageHeader header;
	uint8_t type_data;
} __attribute__((__packed__)) GetCalibration;

typedef struct {
	MessageHeader header;
	uint16_t data[10];
} __attribute__((__packed__)) GetCalibrationReturn;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) SetAccelerationPeriod;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetAccelerationPeriod;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) GetAccelerationPeriodReturn;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) SetMagneticFieldPeriod;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetMagneticFieldPeriod;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) GetMagneticFieldPeriodReturn;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) SetAngularVelocityPeriod;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetAngularVelocityPeriod;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) GetAngularVelocityPeriodReturn;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) SetAllDataPeriod;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetAllDataPeriod;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) GetAllDataPeriodReturn;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) SetOrientationPeriod;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetOrientationPeriod;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) GetOrientationPeriodReturn;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) SetQuaternionPeriod;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetQuaternionPeriod;

typedef struct {
	MessageHeader header;
	uint32_t period;
} __attribute__((__packed__)) GetQuaternionPeriodReturn;

typedef struct {
	MessageHeader header;
	int16_t x;
	int16_t y;
	int16_t z;
} __attribute__((__packed__)) AccelerationSignal;

typedef struct {
	MessageHeader header;
	int16_t x;
	int16_t y;
	int16_t z;
} __attribute__((__packed__)) MagneticFieldSignal;

typedef struct {
	MessageHeader header;
	int16_t x;
	int16_t y;
	int16_t z;
} __attribute__((__packed__)) AngularVelocitySignal;

typedef struct {
	MessageHeader header;
	int16_t acc_x;
	int16_t acc_y;
	int16_t acc_z;
	int16_t mag_x;
	int16_t mag_y;
	int16_t mag_z;
	int16_t ang_x;
	int16_t ang_y;
	int16_t ang_z;
	int16_t temperature;
} __attribute__((__packed__)) AllDataSignal;

typedef struct {
	MessageHeader header;
	int16_t roll;
	int16_t pitch;
	int16_t yaw;
} __attribute__((__packed__)) OrientationSignal;

typedef struct {
	MessageHeader header;
	float x;
	float y;
	float z;
	float w;
} __attribute__((__packed__)) QuaternionSignal;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) OrientationCalculationOn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) OrientationCalculationOff;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) IsOrientationCalculationOn;

typedef struct {
	MessageHeader header;
	bool orientation_calculation_on;
} __attribute__((__packed__)) IsOrientationCalculationOnReturn;

void get_acceleration(const ComType com, const GetAcceleration *data);
void get_magnetic_field(const ComType com, const GetMagneticField *data);
void get_angular_velocity(const ComType com, const GetAngularVelocity *data);
void get_all_data(const ComType com, const GetAllData *data);
void get_orientation(const ComType com, const GetOrientation *data);
void get_quaternion(const ComType com, const GetQuaternion *data);
void get_imu_temperature(const ComType com, const GetIMUTemperature *data);
void leds_on(const ComType com, const LedsOn *data);
void leds_off(const ComType com, const LedsOff *data);
void are_leds_on(const ComType com, const AreLedsOn *data);
void set_acceleration_range(const ComType com, const SetAccelerationRange *data);
void get_acceleration_range(const ComType com, const GetAccelerationRange *data);
void set_magnetometer_range(const ComType com, const SetMagnetometerRange *data);
void get_magnetometer_range(const ComType com, const GetMagnetometerRange *data);
void set_convergence_speed(const ComType com, const SetConvergenceSpeed *data);
void get_convergence_speed(const ComType com, const GetConvergenceSpeed *data);
void set_calibration(const ComType com, const SetCalibration *data);
void get_calibration(const ComType com, const GetCalibration *data);
void set_acceleration_period(const ComType com, const SetAccelerationPeriod *data);
void get_acceleration_period(const ComType com, const GetAccelerationPeriod *data);
void set_magnetic_field_period(const ComType com, const SetMagneticFieldPeriod *data);
void get_magnetic_field_period(const ComType com, const GetMagneticFieldPeriod *data);
void set_angular_velocity_period(const ComType com, const SetAngularVelocityPeriod *data);
void get_angular_velocity_period(const ComType com, const GetAngularVelocityPeriod *data);
void set_all_data_period(const ComType com, const SetAllDataPeriod *data);
void get_all_data_period(const ComType com, const GetAllDataPeriod *data);
void set_orientation_period(const ComType com, const SetOrientationPeriod *data);
void get_orientation_period(const ComType com, const GetOrientationPeriod *data);
void set_quaternion_period(const ComType com, const SetQuaternionPeriod *data);
void get_quaternion_period(const ComType com, const GetQuaternionPeriod *data);
void orientation_calculation_on(const ComType com, const OrientationCalculationOn *data);
void orientation_calculation_off(const ComType com, const OrientationCalculationOff *data);
void is_orientation_calculation_on(const ComType com, const IsOrientationCalculationOn *data);

#endif
